express-notes-api/server.js

```diff
@@ -0,0 +1,157 @@
+import express from 'express';
+import { writeFile, readFile } from 'node:fs/promises';
+
+// --- Reusable functions ---
+function handleError(res, err) {
+  console.error('/api/notes error', err);
+  res.status(500).send({ error: 'An unexpected error occurred.' });
+}
+
+function evaluateId(res, journal, id) {
+  const validIds = Object.keys(journal.notes);
+  const errorMsg = `${id} is not a valid ID. Please provide a valid ID: ${validIds.join(', ')}`;
+  if (isNaN(id) || Number(id) < 0) {
+    res.status(400).send(errorMsg);
+  } else if (!validIds.includes(id)) {
+    res.status(404).send(errorMsg);
+  } else {
+    return true;
+  }
+}
+
+function evaluateReqBody(req, res, action) {
+  let userInputs = Object.keys(req.body);
+  userInputs = userInputs.filter((key) => key !== 'id');
+  switch (action) {
+    case 'delete':
+      if (userInputs.length > 0) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised the above inputs were not be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+    case 'put':
+      if (!userInputs.includes('content') || userInputs.length > 1) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised only notes assigned to "content" will be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+    case 'post':
+      if (!userInputs.includes('content') || userInputs.length > 1) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised only notes assigned to "content" will be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+  }
+}
+
+/** Takes an Express response and user-provided id as inputs.
+ *  If there is an ID, return the entry for that given ID. Otherwise, return all entries. */
+async function loadNotes(res, id) {
+  try {
+    const data = await readFile('data.json', 'utf8');
+    const dataJson = JSON.parse(data);
+    if (id) {
+      if (evaluateId(res, dataJson, id)) {
+        return dataJson.notes[id];
+      }
+    } else {
+      return dataJson;
+    }
+  } catch (err) {
+    handleError(res, err);
+  }
+}
+
+async function uploadNotes(res, input) {
+  try {
```

Comment:
Since you are always calling this inside a try/catch, it's not necessary to have an additional try/catch here. Just let the exception propagate from here up to the try/catch inside the caller. That makes your code simpler since you won't have to check the return value of `uploadNotes` --> either it succeeds or it throws an exception.

The same thing applies to `loadNotes`. Exceptions are awesome, use them! Let the exception indicate that the operation failed, then you don't have to check the result of every function.